### PR TITLE
Combining Inference and PEFT Tokens in a Batch

### PR DESCRIPTION
**Pull request by jiazhihao**
**Description of changes:**

This PR combines LLM inference and PEFT tokens in a batch to enable kernel-level optimizations.

**TODOs (first release):**
- [x] Fix fusion issue
- [x] Optimizers (PR #1290)
- [x] Align cross entropy loss
- [ ] Implement code to initialize a new PEFT from scratch (instead of continuing the finetuning of existing ones)
- [ ] Support upload back to Huggingface (PR #1300)
- [x] Python interface, similar to PEFT
- [x] Add option to add LoRA layers to every linear layer
- [ ] Demo!

**TODOs (remaining ones):**
- [ ] Move QKV and output projections from attention into separate operators, to enable adding LoRA to them (PR #1428).
- [ ] Update implementation, debug and merge token-level finetuning (https://arxiv.org/abs/2402.18789, draft PR #1233)
- [ ] Align models other than LLAMA (opt, falcon, etc...)
- [ ] Add support for LLAMA-3, Phi, Mistral
- [ ] group gemm kernels for multi-tenant lora
- [ ] HIP support
- [x] Add softmax automatically (for categorical cross-entropy loss)

**Related Issues:**

Linked Issues:
- Issue #

Issues closed by this PR:
- Closes #

- - -
This change is [Reviewable](https://reviewable.io/reviews/flexflow/FlexFlow/1153)

### CODE REVIEW

File: inference/models/llama.cc [hunk 1]
```diff
@@ -255,7 +257,8 @@ void LLAMA::create_llama_model(FFModel &ff,
       output = ff.sampling(softmax, generation_config.topp);
     } else {
       // output = ff.arg_top_k(dense, /*k=*/1, false);
-      output = ff.argmax(dense, /*beam_Search*/ false);
+      Tensor softmax = ff.softmax(dense, -1);
```
Comment (goliaro) [inline inference/models/llama.cc:260]: @jiazhihao why is this change required?

Comment (jiazhihao) [reply inference/models/llama.cc:260]: Because we used softmax cross entropy loss, so the last layer of the LLM must be softmax

File: src/runtime/request_manager.cc [hunk 2]
```diff
@@ -378,131 +453,175 @@ BatchConfig RequestManager::prepare_next_batch(BatchConfig const &old_bc,
...
-            request.tokens.at(0) == bos_token_id) {
-          output = "<s> " + output;
-        }
-        {
-          // update generation result
+
+      if (request.req_type == RequestType::REQ_FINETUNING) {
+        // fine-tuning requests don't automatically carry over to the next
+        // batch, we only do so if there is space left after adding new
+        // inference requests
```
Comment (Flechman) [inline src/runtime/request_manager.cc:466]: Based on this comment, and based on how step 3 of `prepare_next_batch` is done (especially its condition to check whether index `i` is free space in the batch), I think `new_bc.request_completed[i]` should be set to `true` here.
Right now it should work without, because `new_bc.request_completed[i]` is not initialized so has a garbage value which has low probability of being 0, and so the condition `if(new_bc.request_completed[i])` in step 3 is evalutated (with high prob) to `true` already. But for consistency I think `new_bc.request_completed[i] = true` should be added here.

Comment (goliaro) [reply src/runtime/request_manager.cc]: Good catch! Let me make the fix!
